contrib/cf-remote/cf_remote/spawn.py

```diff
@@ -0,0 +1,267 @@
+
+    @property
+    def state(self):
+        return self._data.state
+
+    @property
+    def public_ips(self):
+        return self._data.public_ips or []
+
+    @property
+    def private_ips(self):
+        return self._data.private_ips or []
+
+    @property
+    def info(self):
+        ret = {
+            "platform": self.platform,
+            "region": self.region,
+            "size": self.size,
+            "private_ips": self.private_ips,
+            "public_ips": self.public_ips,
+            "uuid": self.uuid,
+        }
+        if self.role:
+            ret["role"] = self.role
+        if self.key_pair:
+            ret["key_pair"] = self.key_pair
+        if self.security_groups:
+            ret["security_groups"] = self.security_groups
+        return ret
+
+    def __str__(self):
+        return "%s: %s" % (self.name, self.info)
+
+    def destroy(self):
+        log.info("Destroying VM '%s'" % self._name)
+        self._node.destroy()
+        self._node = None
+        self._driver = None
+
+
+def _get_unused_name(used_names, prefix, random_suffix_length):
+    random_part = ''.join(random.sample(string.ascii_lowercase, random_suffix_length))
+    name = "%s-%s" % (prefix, random_part)
+    while name in used_names:
+        random_part = ''.join(random.sample(string.ascii_lowercase, random_suffix_length))
+        name = "%s-%s" % (prefix, random_part)
+
+    return name
+
+
+def get_cloud_driver(provider, creds, region):
+    driver_spec = _DriverSpec(provider, creds, region)
+    if driver_spec in _DRIVERS:
+        driver = _DRIVERS[driver_spec]
+    else:
+        cls = get_driver(Provider.EC2)
+        driver = cls(creds.key, creds.secret, region=region)
+
+        # somehow driver.region is always None unless we set it explicitly
+        driver.region = region
+
+        _DRIVERS[driver_spec] = driver
+
+    return driver
+
+
+def spawn_vm_in_aws(platform, aws_creds, key_pair, security_groups, region, name=None, size=None, role=None):
+    driver = get_cloud_driver(Providers.AWS, aws_creds, region)
+    existing_vms = driver.list_nodes()
+
+    if name is None:
+        name = _get_unused_name([vm.name for vm in existing_vms], platform, _NAME_RANDOM_PART_LENGTH)
+    else:
+        if any(vm.state in (0, 'running') and vm.name == name for vm in existing_vms):
+            raise ValueError("VM with the name '%s' already exists" % name)
+
+    aws_platform = aws_platforms[platform]
+    size = size or aws_platform["size"]
+    ami = aws_platform["ami"]
+
+    log.info("Spawning new '%s' VM in AWS (AMI: %s, size=%s)" % (platform, ami, size))
+    node = driver.create_node(
+        name=name,
+        image=NodeImage(id=ami, name=None, driver=driver),
+        size=NodeSize(id=size, name=None, ram=None,
+		      disk=None, bandwidth=None, price=None, driver=driver),
+        ex_keyname=key_pair,
+        ex_security_groups=security_groups,
```

Comment:
We should tag instances, for example `spawned-by-cf-remote` and the result of `whoami`. Can be a separate feature / PR.

Reply:
+1